apps/web/public/sw.js

```diff
@@ -0,0 +1,411 @@
+      ),
+    ]);
+  } catch (err) {
+    console.log(err);
+    throw err;
+  }
+}
+
+// this is written to avoid callback failure for first time users as the redirection fails only in the tab they used to login
+// goes away with hard reload (CTRL + SHIFT + R)
+const isAuth = (url) => {
+  if (url.includes("/api/auth/signout")) clearAllData();
+  return (
+    url.includes("welcome") ||
+    url.includes("api/auth") ||
+    url.includes("newsletters")
+  );
+};
+
+const loadLocalMail = async (decreasing = false) => {
+  if (!DB) await openDB();
+
+  const data = await getLocalMail(decreasing);
+  if (data.length > 0) return data;
+  return null;
+};
+
+const getLocalMail = (decreasing, keyRange) => {
+  return new Promise((resolve, reject) => {
+    const tx = DB.transaction(EMAIL_STORE.name, "readonly");
+    const store = tx.objectStore(EMAIL_STORE.name);
+    const index = store.index("timestampIDX");
+
+    // this request will get a range query with increasing order of timestamp (i.e. oldest to latest)
+    // if keyRange is undefined, it will get all the data.
+    const req = index.getAll(keyRange);
+
+    req.onerror = () => resolve(req.error);
+    req.onsuccess = () =>
+      resolve(decreasing ? req.result : req.result.reverse());
+  });
+  // TODO: make a utility function for creating transaction.
+};
+
+const saveMails = async (mails) => {
+  if (!DB) {
+    await openDB();
+  }
+  const tx = DB.transaction(EMAIL_STORE.name, "readwrite");
+  const store = tx.objectStore(EMAIL_STORE.name);
+
+  for (const mail of mails.emails) {
+    const req = store.put(mail);
+    req.onsuccess = console.log("added mail");
+    req.onerror = (e) =>
+      console.log("error occurred while adding mails to index", e);
+  }
+};
+
+async function fetchAndUpdateMails(request, decreasing) {
+  const resp = await fetch(request);
+  const restMails = await resp.json();
+  await saveMails(restMails);
+  const allMails = await loadLocalMail(decreasing);
+  messageClients(MESSAGES.MAILS_UPDATED, allMails);
+}
+
+/* Analytics/ Stats Part */
+
+function getStatsByPeriod(options) {
+  return new Promise(async (resolve, reject) => {
+    const { period, fromDate, toDate } = options;
+    // TODO: Figure out a way to use import statment in service worker without causing error, so you can use lodash in service worker ->Uncaught SyntaxError: Cannot use import statement outside a module (at service-worker.js:1:1)
+
+    const keyRange = IDBKeyRange.bound(+fromDate, +toDate, true, true);
+
+    const localData = await getLocalMail(false, keyRange);
+
+    resolve(getAllMailsClusterdbyPeriod(+fromDate, +toDate, period, localData));
+  });
+}
+
+// all inclusive
+const numberOfClusters = (start, end, size) => {
+  const window = (end - start + 1) / size;
+  return Math.floor(window) + (window - Math.floor(window) > 0 ? 1 : 0);
+};
+
+const getClusteredArray = (start, end, size) => {
+  const nClusters = numberOfClusters(start, end, size);
+  const clusters = [];
```

Comment:
A cluster is a period of time with all the stats about all activity during that period. It could be of size day/week/month/year.